File: tuan05/cai_tui.py
def cai_tui(n, W, w, v):
    f = [[0] * (W + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, W + 1):
            if j < w[i - 1]:
                f[i][j] = f[i - 1][j]
            else:
                f[i][j] = max(f[i - 1][j], f[i - 1][j - w[i - 1]] + v[i - 1])

    print("Bang f:")
    for row in f:
        print(row)

    print("Gia tri lon nhat:", f[n][W])

    # Truy vet
    i, j = n, W
    chon = []
    ten_do_vat = ["A", "B", "C", "D", "E"]
    while i > 0 and j > 0:
        if f[i][j] != f[i - 1][j]:
            chon.append(ten_do_vat[i - 1])
            j -= w[i - 1]
        i -= 1
    chon.reverse()
    print("Tap do vat duoc chon:", chon)


n = 5
W = 11
w = [2, 3, 4, 5, 7]
v = [3, 7, 9, 12, 16]
cai_tui(n, W, w, v)

File: tuan05/cai_tui.jl
function cai_tui(n, W, w, v)
    f = zeros(Int, n + 1, W + 1)
    
    for i in 1:n
        for j in 1:W
            if j < w[i]
                f[i + 1, j + 1] = f[i, j + 1]
            else
                f[i + 1, j + 1] = max(f[i, j + 1], f[i, j - w[i] + 1] + v[i])
            end
        end
    end

    println("Bang f:")
    display(f)
    println("\nGia tri lon nhat: ", f[n + 1, W + 1])

    i, j = n, W
    chon = String[]
    ten_do_vat = ["A", "B", "C", "D", "E"]
    while i > 0 && j > 0
        if f[i + 1, j + 1] != f[i, j + 1]
            push!(chon, ten_do_vat[i])
            j -= w[i]
        end
        i -= 1
    end
    reverse!(chon)
    println("Tap do vat duoc chon: ", chon)
end

n = 5
W = 11
w = [2, 3, 4, 5, 7]
v = [3, 7, 9, 12, 16]
cai_tui(n, W, w, v)